---
title: spliceQTL
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

This vignette requires both a local machine and the BBMRI virtual machine. On the virtual machine, execute this chunk to set the library path, update the R package spliceQTL, and set the working directory.

```{r virtual machine,eval=FALSE}
lib <- "/virdir/Scratch/arauschenberger/library"
.libPaths(lib)
devtools::install_github("rauschenberger/spliceQTL",lib=lib)
library("spliceQTL",lib.loc=lib)
path <- "/virdir/Scratch/arauschenberger/spliceQTL"
setwd(path)
```

On a local machine with PLINK, execute this chunk to obtain the Gevuadis SNP data. Then move the files from the local to the virtual machine.

```{r local data,eval=FALSE}
for(chr in 1:22){ 
    spliceQTL::get.snps.geuvadis(chr=chr,data="N:/semisup/data/eQTLs",
                                 path="N:/spliceQTL/data/Geuvadis")
}
```

On the virtual machine, execute this chunk to obtain the BBMRI SNP data, the Geuvadis exon data, and the BBMRI exon data. Choose one out of the six biobanks (CODAM, LL, LLS, NTR, PAN, RS).

```{r remote data,eval=FALSE}
for(chr in 1:22){
    spliceQTL::get.snps.bbmri(chr=chr,biobank="LLS",path=path,size=500*10^3)
}
spliceQTL::get.exons.geuvadis(path=path)
spliceQTL::get.exons.bbmri(path=path)
```

On the virtual machine, execute this chunk to test for alternative splicing. See the documentation of the R package spliceQTL for further information.

```{r analysis,eval=FALSE}
for(chr in 1:22){
  for(data in c("Geuvadis","LLS")){
    set.seed(1)
    
    cat("Analysing",data,"data:",as.character(Sys.time()),"\n")
    if(data=="Geuvadis"){
      load(file.path(path,"Geuvadis.exons.RData"),verbose=TRUE)
    } else {
      load(file.path(path,"BBMRI.exons.RData"),verbose=TRUE)
      cond <- sapply(strsplit(x=rownames(exons),split=":"),function(x) x[[1]]==data)
      exons <- exons[cond,]
    }
    load(file.path(path,paste0(data,".chr",chr,".RData")),verbose=TRUE)
    
    cat("Matching samples:","\n")
    list <- spliceQTL::match.samples(exons,snps)
    exons <- list$exons; snps <- list$snps; rm(list)
    
    cat("Adjusting samples:","\n")
    exons <- spliceQTL::adjust.samples(x=exons) # slow!
    exons <- asinh(x=exons)
    
    cat("Adjusting covariates:","\n")
    names <- strsplit(x=colnames(exons),split="_") # exon names
    length <- sapply(names,function(x) as.integer(x[[3]])-as.integer(x[[2]])) # exon length
    exons <- spliceQTL::adjust.covariates(x=exons,group=gene_id,offset=length) # slow!
    
    cat("Mapping exons:","\n")
    map <- list()
    map$genes <- spliceQTL::map.genes(chr=chr,path=path)
    map$exons <- spliceQTL::map.exons(gene=as.character(map$genes$gene_id),exon=gene_id)
    
    cat("Mapping SNPs:","\n")
    names <- strsplit(x=colnames(snps),split=":")
    snp.chr <- sapply(names,function(x) x[[1]])
    snp.pos <- sapply(names,function(x) x[[2]])
    map$snps <- spliceQTL::map.snps(gene.chr=map$genes$chr,
                                    gene.start=map$genes$start,
                                    gene.end=map$genes$end,
                                    snp.chr=snp.chr,snp.pos=snp.pos)
    
    cat("Dropping genes:","\n")
    map <- spliceQTL::drop.trivial(map=map)
    
    cat("Testing:",as.character(Sys.time())," -> ")
    pvalue <- spliceQTL::test.multiple(Y=exons,X=snps,map=map,rho=c(0,0.5),spec=4)
    cat(as.character(Sys.time()),"\n")
    
    save(object=pvalue,file=file.path(path,paste0("pval.",data,".chr",chr,".RData")))
  }
}
```

On the virtual machine, execute this chunk to compare the results between the Geuvadis and the BBMRI project.

```{r,eval=FALSE}
cor <- chisq <- rep(NA,length=22)
for(chr in 1:22){
    load(file.path(path,"pval.Geuvadis.chr",chr,".RData"))
    a <- pvalue; pvalue <- NA
    load(file.path(path,"pval.LLS.chr",chr,".RData"))
    b <- pvalue; pvalue <- NA

    names <- intersect(rownames(a),rownames(b))
    cor[chr] <- stats::cor(a[names,1],b[names,1],method="spearman")

    a <- stats::p.adjust(a[names,1])<0.05
    b <- stats::p.adjust(b[names,1])<0.05
    chisq[chr] <- stats::chisq.test(table(a,b))
}
```


